core/remote/errors.js

```diff
@@ -171,42 +188,111 @@ const wrapError = (err /*: FetchError |  Error */) /*: RemoteError */ => {
           err
         })
       case 412:
-        return new RemoteError({
-          code: NEEDS_REMOTE_MERGE_CODE,
-          message: 'Known remote document revision is outdated',
-          err
-        })
+        if (sourceParameter(err) === 'If-Match') {
+          // Revision error
+          return new RemoteError({
+            code: NEEDS_REMOTE_MERGE_CODE,
+            message: 'The known remote document revision is outdated',
+            err
+          })
+        } else if (sourceParameter(err) === 'dir-id') {
+          // The directory is asked to move to one of its sub-directories
+          return new RemoteError({
+            code: INVALID_FOLDER_MOVE_CODE,
+            message:
+              'The folder would be moved wihtin one of its sub-folders on the remote Cozy',
+            err
+          })
+        } else {
+          // Invalid hash or content length error
+          return new RemoteError({
+            code: INVALID_METADATA_CODE,
+            message: 'The local metadata for the document is corrupted',
+            err
+          })
+        }
       case 413:
         return new RemoteError({
           code: NO_COZY_SPACE_CODE,
           message: 'Not enough space available on remote Cozy',
           err
         })
+      case 422:
+        if (sourceParameter(err) === 'name') {
+          return new RemoteError({
+            code: INVALID_NAME_CODE,
+            message:
+              'The name of the document contains characters forbidden by the remote Cozy',
+            err
+          })
+        } else if (sourceParameter(err) === 'path') {
+          return new RemoteError({
+            code: PATH_TOO_DEEP_CODE,
+            message:
+              'The path of the document has too many levels for the remote Cozy',
+            err
+          })
+        } else {
+          return new RemoteError({
+            code: INVALID_METADATA_CODE,
+            message: 'The local metadata for the document is corrupted',
+            err
+          })
+        }
       default:
-        // TODO: Merge with UnreachableError?!
-        return new RemoteError({
-          code: UNREACHABLE_COZY_CODE,
-          message: 'Cannot reach remote Cozy',
-          err
-        })
+        if (status > 400 && status < 500) {
+          return new RemoteError({
+            code: UNKNOWN_INVALID_DATA_ERROR_CODE,
+            message:
+              'The data sent to the remote Cozy is invalid for some unhandled reason',
+            err
+          })
+        } else {
+          // TODO: Merge with UnreachableError?!
```

Comment:
`TODO` spotted

Reply:
Yes indeed. But `UnreachableError` is only used when opening a Cozy Note and using a `RemoteError` instead means I also have to deal with `CozyDocumentMissingError` which should not be in the remote errors module…
I think this PR is large enough and we need to get this code out to fix users' problems so I'll deal with the refactoring later :).